pyro/contrib/funsor/infer/traceenum_elbo.py

```diff
@@ -39,9 +42,70 @@ def terms_from_trace(tr):
+                enum(first_available_dim=(-self.max_plate_nesting-1) if self.max_plate_nesting else None):
+            guide_tr = trace(guide).get_trace(*args, **kwargs)
+            model_tr = trace(replay(model, trace=guide_tr)).get_trace(*args, **kwargs)
+
+        # extract from traces all metadata that we will need to compute the elbo
+        guide_terms = terms_from_trace(guide_tr)
+        model_terms = terms_from_trace(model_tr)
+
+        # build up a lazy expression for the elbo
+        with funsor.interpreter.interpretation(funsor.terms.lazy):
+            # identify and contract out auxiliary variables in the model with partial_sum_product
+            contracted_factors, uncontracted_factors = [], []
+            for f in model_terms["log_factors"]:
+                if model_terms["measure_vars"].intersection(f.inputs):
+                    contracted_factors.append(f)
+                else:
+                    uncontracted_factors.append(f)
+            # incorporate the effects of subsampling and handlers.scale through a common scale factor
+            contracted_costs = [model_terms["scale"] * f for f in funsor.sum_product.modified_partial_sum_product(
+                funsor.ops.logaddexp, funsor.ops.add,
+                model_terms["log_measures"] + contracted_factors,
+                plate_to_step=model_terms["plate_to_step"],
+                eliminate=model_terms["measure_vars"] | frozenset({
+                    p for p, s in model_terms["plate_to_step"].items() if s})
+            )]
+
+            costs = contracted_costs + uncontracted_factors  # model costs: logp
+            costs += [-f for f in guide_terms["log_factors"]]  # guide costs: -logq
+
+            # finally, integrate out guide variables in the elbo and all plates
+            # dict1 | dict2 in python 3.9
+            plate_to_step = {**guide_terms["plate_to_step"], **model_terms["plate_to_step"]}
+            plate_vars = frozenset(plate_to_step.keys())
+            elbo = to_funsor(0, output=funsor.Real)
+            for cost in costs:
+                # compute the marginal logq in the guide corresponding to this cost term
+                factors = funsor.sum_product.modified_partial_sum_product(
+                    funsor.ops.logaddexp, funsor.ops.add,
+                    guide_terms["log_measures"],
+                    plate_to_step=plate_to_step,
+                    eliminate=(plate_vars | guide_terms["measure_vars"]) - frozenset(cost.inputs)
+                )
+                log_prob = reduce(funsor.ops.add, factors, funsor.Number(funsor.ops.UNITS[funsor.ops.add]))
+                # compute the expected cost term E_q[logp] or E_q[-logq] using the marginal logq for q
+                elbo_term = funsor.Integrate(log_prob, cost, guide_terms["measure_vars"] & frozenset(cost.inputs))
+                elbo += elbo_term.reduce(funsor.ops.add, plate_vars & frozenset(cost.inputs))
```

Comment:
Note these two lines will not be correct for markov plates; you should try to write a guide enumeration test where this causes the sequential and vectorized markov ELBO values to disagree. I believe some of the test cases I suggested would be sufficient.

You might need to rewrite this as another `modified_partial_sum_product` computation. Note that 
```py
funsor.Integrate(log_prob, integrand, reduced_vars)
```
is equivalent to
```py
(log_prob.exp() * integrand).reduce(ops.add, reduced_vars)
```

Reply:
> Note these two lines will not be correct for markov plates

Ok. I haven't checked the math yet. So there might be some more errors.

> is equivalent to

Thanks for explaining this.

Reply:
After taking a closer look it seems like `modified_partial_sum_product` might not work for guide enumerated models. The reason is that cost (integrand) `logq(z)` of transition factors has to be split into the sum factors `logq(z_0) + logq(z_1|z_0) + ...` before integrating w.r.t. `q(z_0)q(z_1|z_0) ...`. I can explain this in more details at the Zoom meeting.

Reply:
In other words the `prod` operation in the integrand is `ops.add` and in the `log_prob.exp()` is `ops.mul`.